File: THCS/2022_2023/BenTre_2223/Bai2/BAI2.py
def can_divide_into_groups(n, acquaintance_matrix):
    group = [-1] * n  # -1: chưa được phân nhóm, 0: nhóm 1, 1: nhóm 2

    def dfs(person, current_group):
        group[person] = current_group
        next_group = 1 - current_group  # Chuyển sang nhóm khác

        for acquaintance in range(n):
            if acquaintance_matrix[person][acquaintance] == 1:
                if group[acquaintance] == -1:  # Nếu chưa được phân nhóm
                    if not dfs(acquaintance, next_group):
                        return False
                elif group[acquaintance] == current_group:  # Nếu cùng nhóm
                    return False
        return True

    for person in range(n):
        if group[person] == -1:  # Nếu chưa được phân nhóm
            if not dfs(person, 0):
                return False, []

    # Tạo danh sách cho từng nhóm
    group1 = [i + 1 for i in range(n) if group[i] == 0]
    group2 = [i + 1 for i in range(n) if group[i] == 1]

    return True, (group1, group2)


# Đọc dữ liệu từ file CHIANHOM.INP
with open("CHIANHOM.INP", "r") as infile:
    n = int(infile.readline().strip())
    acquaintance_matrix = [list(map(int, infile.readline().strip().split())) for _ in range(n)]

# Kiểm tra và lấy kết quả
can_divide, groups = can_divide_into_groups(n, acquaintance_matrix)

# Ghi kết quả ra file CHIANHOM.OUT
with open("CHIANHOM.OUT", "w") as outfile:
    if not can_divide:
        outfile.write("0\n")
    else:
        outfile.write(" ".join(map(str, groups[0])) + "\n")
        outfile.write(" ".join(map(str, groups[1])) + "\n")

File: THCS/2022_2023/BenTre_2223/Bai2/BAI2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>

using namespace std;

bool canDivideIntoGroups(int n, const vector<vector<int>>& acquaintanceMatrix, vector<int>& group) {
    for (int person = 0; person < n; ++person) {
        if (group[person] == -1) {  // Nếu chưa được phân nhóm
            queue<int> q;
            q.push(person);
            group[person] = 0;  // Gán nhóm 0

            while (!q.empty()) {
                int current = q.front();
                q.pop();
                int nextGroup = 1 - group[current];  // Chuyển sang nhóm khác

                for (int acquaintance = 0; acquaintance < n; ++acquaintance) {
                    if (acquaintanceMatrix[current][acquaintance] == 1) { // Nếu là quen biết
                        if (group[acquaintance] == -1) { // Chưa phân nhóm
                            group[acquaintance] = nextGroup;
                            q.push(acquaintance);
                        } else if (group[acquaintance] == group[current]) { // Cùng nhóm
                            return false; // Không thể chia nhóm
                        }
                    }
                }
            }
        }
    }
    return true;
}

int main() {
    ifstream infile("CHIANHOM.INP");
    ofstream outfile("CHIANHOM.OUT");

    int n;
    infile >> n;

    vector<vector<int>> acquaintanceMatrix(n, vector<int>(n));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            infile >> acquaintanceMatrix[i][j];
        }
    }

    vector<int> group(n, -1); // -1: chưa phân nhóm, 0: nhóm 1, 1: nhóm 2
    if (!canDivideIntoGroups(n, acquaintanceMatrix, group)) {
        outfile << "0\n"; // Không thể chia nhóm
    } else {
        vector<int> group1, group2;
        for (int i = 0; i < n; ++i) {
            if (group[i] == 0) {
                group1.push_back(i + 1); // Thêm vào nhóm 1 (bắt đầu từ 1)
            } else {
                group2.push_back(i + 1); // Thêm vào nhóm 2
            }
        }
        for (int person : group1) {
            outfile << person << " ";
        }
        outfile << "\n";
        for (int person : group2) {
            outfile << person << " ";
        }
        outfile << "\n";
    }

    infile.close();
    outfile.close();
    return 0;
}
